File: clojure/run-length-encoding/src/run_length_encoding.clj
(ns run-length-encoding
  (:require [clojure.string :as string]))

(defn- encode-match [[sequence character]]
  (str (count sequence) character))

(defn- decode-match [[_ run-length character]]
  (string/join (repeat (Integer/parseInt run-length) character)))

(defn run-length-encode
  "encodes a string with run-length-encoding"
  [plain-text]
  (string/replace plain-text #"(\D)\1+" encode-match))

(defn run-length-decode
  "decodes a run-length-encoded string"
  [cipher-text]
  (string/replace cipher-text #"(\d+)(\D)" decode-match))

File: javascript/run-length-encoding/run-length-encoding.js
export function encode(str) {
    return str.replace(/([a-zA-Z ])(\1+)/g, (match, first) => `${match.length}${first}`);
}

export function decode(str) {
    return str.replace(/([0-9]+)([a-zA-Z ])/g, (_, count, char) => char.repeat(count));
}
